numpy/core/src/multiarray/temp_avoid.c

```diff
@@ -0,0 +1,367 @@
+ *
+ *        0 LOAD_FAST                0 (a)
+ *        3 LOAD_FAST                1 (b)
+ *        6 BINARY_ADD
+ *        7 LOAD_FAST                1 (b)
+ *       10 BINARY_ADD
+ *
+ * The two named variables get their reference count increased by the load
+ * instructions so they always have a reference count larger than 1.
+ * The temporary of the first BINARY_ADD on the other hand only has a count of
+ * 1. Only temporaries can have a count of 1 in python so we can use this to
+ * transform the second operation into an inplace operation and not affect the
+ * output of the program.
+ * CPython does the same thing to resize memory instead of copying when doing
+ * string concatenation.
+ *
+ * A complication is that a C program could call the PyNumber functions
+ * directly on arrays with reference count of 1. This is handled by checking
+ * the call stack to verify that we have been called directly from the cpython
+ * interpreter.
+ * This is done by checking that all functions in the callstack until the frame
+ * evaluation function are located in cpython or numpy.
+ * This is an expensive operation so temporaries are only avoided for rather
+ * large arrays.
+ *
+ * A possible future improvement would be to change cpython to give as access
+ * to the top of the stack. Then we could just check that the objects involved
+ * are on the cpython stack instead of checking the function callstack.
+ */
+
+#if defined HAVE_BACKTRACE && defined HAVE_DLFCN_H && ! defined PYPY_VERSION
+/* 1 prints elided operations, 2 prints stacktraces */
+#define NPY_ELIDE_DEBUG 0
+#define NPY_MAX_STACKSIZE 10
+
+#if PY_VERSION_HEX >= 0x03060000
+/* TODO can pep523 be used to somehow? */
+#define PYFRAMEEVAL_FUNC "_PyEval_EvalFrameDefault"
+#else
+#define PYFRAMEEVAL_FUNC "PyEval_EvalFrameEx"
+#endif
+/*
+ * Heuristic number at which backtrace overhead generation becomes less than
+ * speed gain. Depends on stack depth being checked.
+ * Measurements with 10 stacks show it getting worthwhile around 100KiB but to
+ * be conservative put it higher around where the L2 cache spills.
+ */
+#ifndef Py_DEBUG
+#define NPY_MIN_ELIDE_BYTES (256 * 1024)
+#else
+/*
+ * in debug mode always elide but skip scalars as these can convert to 0d array
+ * during in place operations
+ */
+#define NPY_MIN_ELIDE_BYTES (32)
+#endif
+#include <dlfcn.h>
+#include <execinfo.h>
+
+/*
+ * linear search pointer in table
+ * number of pointers is usually quite small but if a performance impact can be
+ * measured this could be converted to a binary search
+ */
+static int
+find_addr(void * addresses[], npy_intp naddr, void * addr)
+{
+    npy_intp j;
+    for (j = 0; j < naddr; j++) {
+        if (addr == addresses[j]) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static int
+check_callers(int * cannot)
+{
+    /*
+     * get base addresses of multiarray, python, check if
+     * backtrace is in these libraries only calling dladdr if a new max address
```

Comment:
the assumption is that a library is loaded linear into the address space which is at least the case on linux and probably all other platforms. Then we can just check the upper and lower bound of the addresses to avoid expensive dladdr calls.